File: interfaces/matlab/examples/brachistochroneAuxdata.m
%-------------- Classical Brachistochrone Problem with Auxdata -----------%
% This example is identitical to brachistochrone.m, but                   %
% the OCP includes axmax,aymax as auxdata, in order to allow to run       %
% several simulations with different sets of data without the need to     %
% generate the MEX function again.                                        %
%-------------------------------------------------------------------------%

clc, clear

if ispc
    addpath('./../../bin');
else
    addpath('./../../lib');
end

% import casadi
import casadi.*

% Solving options
name = 'brachistochroneAuxdata'; % name of the problem
scheme = 'trapz'; %integration scheme: trapz, rk4, euler
skip_hessian = false; % use approximated hessian
N = 500; % number of mesh points
ti = 0; % initial time
tf = 1; % final time

% Data
% g = 10; % now this is an auxdata

% Boundary conditions
xi = [0; 0; 0]; % initial x,y,v
xf = [2; 2]; % final x,y

% Bounds
lbx = [0; 0; -50]; % lower bound for state
ubx = [10; 10; 50]; % upper bound for state
lbu = -pi/2; % lower bound for control
ubu = pi/2; % upper bound for control
lbp = 0; % lower bound for parameter
ubp = 2; % upper bound for parameter
lbc = []; % lower bound for path constraint
ubc = []; % upper bound for path constraint
lbb = [xi; xf]; % lower bound for boundary conditions
ubb = [xi; xf]; % upper bound for boundary conditions
lbq = []; % lower bound for integral constraints
ubq = []; % upper bound for integral constraints

% Create CASADI variables
nx = 3; nu = 1; np = 1; na = 1; % state, control, parameter, and auxdata dimensions
t = SX.sym('t'); % independent variable
x = SX.sym('x', nx); % state
u = SX.sym('u', nu); % control
p = SX.sym('p', np); % parameter
x0 = SX.sym('x0', nx); % initial state
u0 = SX.sym('u0', nu); % initial control
xn = SX.sym('xn', nx); % final state
un = SX.sym('un', nu); % final control
auxdata = SX.sym('auxdata', na); % auxdata

% Model equations ( using time scaled by Tf=p(1) )
g = auxdata(1); % g is an auxdata
xdot = p(1)*[x(3)*sin(u(1));
             x(3)*cos(u(1));
             g*cos(u(1))]; % state derivative
c = []; % path constraint
q = []; % integral constraint
l = 0; % running (Lagrange) cost
m = p(1); % boundary (Mayer) cost
% equivalently, we could also use:
% l = p(1);
% m = 0;

% Boundary conditions
b = [x0;
     xn(1:2)];

% Integrator
h = SX.sym('h'); % time step
f = Function('f', {t, x, u, p, auxdata}, {xdot, l, q}); % ocp dynamics [xdot,l,q] = f(t,x,u,p,auxdata)
switch scheme
    case 'euler' % Euler integration
        x1 = SX.sym('x1', nx); % initial state of interval 
        x2 = SX.sym('x2', nx); % final state of interval 
        [f1, dl1, dq1] = f(t, x1, u, p, auxdata);
        xnext = x1 + h*f1; % euler 
        dl = h*dl1; % euler 
        dx = xnext - x2;
        dq = h*dq1;
    case 'trapz'
        x1 = SX.sym('x1', nx); % initial state of interval 
        x2 = SX.sym('x2', nx); % final state of interval 
        [f1, dl1, dq1] = f(t,   x1, u, p, auxdata); 
        [f2, dl2, dq2] = f(t+h, x2, u, p, auxdata); 
        xnext = x1 + 1/2 * h * (f1+f2);
        dx = xnext - x2;
        dl = 1/2 * h * (dl1+dl2);
        dq = 1/2 * h * (dq1+dq2);
    case 'rk4' % Fixed step Runge-Kutta 4 integrator
        M = 4; % RK4 steps per interval
        x1 = SX.sym('x1', nx); % initial state of interval 
        x2 = SX.sym('x2', nx); % final state of interval 
        xnext = x1; % init
        dl = 0; % init
        dq = 0; % init
        for j=1:M
           [k1, k1_l, k1_q] = f(t, xnext, u, p, auxdata);
           [k2, k2_l, k2_q] = f(t + h/M/2, xnext + h/M/2 * k1, u, p, auxdata);
           [k3, k3_l, k3_q] = f(t + h/M/2, xnext + h/M/2 * k2, u, p, auxdata);
           [k4, k4_l, k4_q] = f(t + h/M, xnext + h/M * k3, u, p, auxdata);
           xnext = xnext + h/M/6*(k1 +2*k2 +2*k3 +k4);
           dl = dl + h/M/6*(k1_l + 2*k2_l + 2*k3_l + k4_l);
           dq = dq + h/M/6*(k1_q + 2*k2_q + 2*k3_q + k4_q);
        end
        dx = xnext - x2;
end

% OCP functions
% In this case the OCP functions have also 'auxdata' as the last argument.
% OCP runnig cost function dl = ocp_runcost(t, x1, u, x2, p, h, auxdata)
% t: time at step k
% x1: state at step k
% x2: state at step k+1
% u: control over the interval [k,k+1]
% p: scalar parameter
% h: time step
% auxdata: auxdata
% dl: integral (i.e. Lagrange) cost over the interval [k,k+1]
ocp_runcost = Function('ocp_runcost', {t, x1, u, x2, p, h, auxdata}, {dl});

% OCP boundary cost function m = ocp_bcscost(x0, u0, xn, un, p, auxdata)
% x0: initial state
% u0: initial control
% xn: final state
% un: final control
% p: scalar parameter
% auxdata: auxdata
% m: boundary (i.e. Mayer) cost
ocp_bcscost = Function('ocp_bcscost', {x0, u0, xn, un, p, auxdata}, {m});

% OCP dynamic function dx = ocp_dyn(t, x1, u, x2, p, h, auxdata)
% t: time at step k
% x1: state at step k
% x2: state at step k+1
% u: control over the interval [k,k+1]
% p: scalar parameter
% h: time step
% auxdata: auxdata
% dx: rhs of integrator over the interval [k,k+1] (which must be dx=0 for
% continuity)
ocp_dyn = Function('ocp_dyn', {t, x1, u, x2, p, h, auxdata}, {dx});

% OCP path function c = ocp_path(t, x, u, p, auxdata)
% t: time at step k
% x: state at step k
% u: control over the interval [k,k+1]
% p: scalar parameter
% auxdata: auxdata
% c: path constraint at step k
ocp_path = Function('ocp_path', {t, x, u, p, auxdata}, {c});

% OCP boundary condition function b = ocp_bcs(x0, u0, xn, un, p, auxdata)
% x0: initial state
% u0: initial control
% xn: final state
% un: final control
% p: scalar parameter
% auxdata: auxdata
% b: boundary condtions
ocp_bcs = Function('ocp_bcs', {x0, u0, xn, un, p, auxdata}, {b});

% OCP integral constraint function dq = ocp_int(t, x1, u, x2, p, h)
% t: time at step k
% x1: state at step k
% x2: state at step k+1
% u: control over the interval [k,k+1]
% p: scalar parameter
% h: time step
% auxdata: auxdata
% dq: integral constraint over the interval [k,k+1]
ocp_int = Function('ocp_int', {t, x1, u, x2, p, h, auxdata}, {dq});

% Build OCP
buildOCP(name, ...
        ocp_runcost, ocp_bcscost, ocp_dyn, ocp_path, ocp_bcs, ocp_int, ...
        skip_hessian);

% Create guess
x0 = [linspace(xi(1),xf(1),N)
      linspace(xi(2),xf(2),N)
      linspace(0,0,N)];
u0 = linspace(0,1,N);
p0 = 1;

% Create problem structure
problem.name = name;
problem.N = N;
problem.ti = ti;
problem.tf = tf;
problem.guess.x = x0;
problem.guess.u = u0;
problem.guess.p = p0;
problem.bounds.lbx = lbx; problem.bounds.ubx = ubx;
problem.bounds.lbp = lbp; problem.bounds.ubp = ubp;
problem.bounds.lbu = lbu; problem.bounds.ubu = ubu;
problem.bounds.lbc = lbc; problem.bounds.ubc = ubc;
problem.bounds.lbb = lbb; problem.bounds.ubb = ubb;
problem.bounds.lbq = lbq; problem.bounds.ubq = ubq;
problem.auxdata = 10; % set also auxdata

% Solve with auxdata = 10
sol1 = minosMex(problem);

% Change auxdata, and solve again starting from previous solution
problem = sol1.next_problem;
problem.auxdata = 5;
sol2 = minosMex(problem);

clear minosMex

% Plot the solution
figure
plot(sol1.t, sol1.x, '-', sol2.t, sol2.x, '--');
xlabel('t'), ylabel('state');
legend('x1','y1', 'v1','x2','y2', 'v2')

figure
plot(sol1.t, sol1.u, '-' ,sol2.t, sol2.u, '--');
xlabel('t'), ylabel('control');
legend('u1','u2')

figure
plot(sol1.x(1,:), sol1.x(2,:), '-', sol2.x(1,:), sol2.x(2,:), '--');
xlabel('x'), ylabel('y');
legend('1','2')
axis equal
set(gca,'YDir','reverse');

File: interfaces/python/examples/brachistochrone_auxdata.py
#!/usr/bin/env python

# Brachistochrone problem with auxdata
import sys
sys.path.append("./../../bin")

import casadi as ca # import CASADI
import minosPy # import MINOS
import matplotlib.pyplot as plt

# Solver options
name = "brachistochroneAuxdata"
skip_hessian = False # use approximated hessian
N = 500
ti = 0
tf = 1

# Data
# g = 10 # now this is an auxdata (set later)

# Create CASADI variables
nx = 3
nu = 1
np = 1
na = 1
t = ca.SX.sym('t') # independent variable
x = ca.SX.sym('x', nx) # state
u = ca.SX.sym('u', nu) # control
p = ca.SX.sym('p', np) # parameter
xi = ca.SX.sym('xi', nx) # initial state
ui = ca.SX.sym('ui', nu) # initial control
xf = ca.SX.sym('xf', nx) # final state
uf = ca.SX.sym('uf', nu) # final control
auxdata = ca.SX.sym('auxdata', na) # auxdata

# Model equations ( using time scaled by Tf=p[0] )
g = auxdata[0]
xdot = ca.vertcat(p[0]*x[2]*ca.sin(u[0]), p[0]*x[2]*ca.cos(u[0]), p[0]*g*ca.cos(u[0])) # state derivative
c = [] # path constraint
q = [] # integral constraint
l = 0 # running cost
m = p[0] # boundary cost

# Boundary conditions
b = ca.vertcat(xi[0], xi[1], xi[2], xf[0], xf[1])

# Integrator (trapezoidal)
h = ca.SX.sym('h')
f = ca.Function('f', [t, x, u, p, auxdata], [xdot, l, q])
x1 = ca.SX.sym('x1', nx) # initial state of interval 
x2 = ca.SX.sym('x2', nx) # final state of interval 
f1, dl1, dq1 = f(t,   x1, u, p, auxdata)
f2, dl2, dq2 = f(t+h, x2, u, p, auxdata)
xnext = x1 + 0.5*h*(f1+f2)
dx = xnext - x2
dl = 0.5*h*(dl1+dl2)
dq = 0.5*h*(dq1+dq2)

# OCP functions
# In this case the OCP functions have also 'auxdata' as the last argument.
# OCP running cost
ocp_runcost = ca.Function('ocp_runcost', [t, x1, u, x2, p, h, auxdata], [dl])
# OCP boundary cost
ocp_bcscost = ca.Function('ocp_bcscost', [xi, ui, xf, uf, p, auxdata], [m])
# OCP dynamic
ocp_dyn = ca.Function('ocp_dyn', [t, x1, u, x2, p, h, auxdata], [dx])
# OCP path
ocp_path = ca.Function('ocp_path', [t, x, u, p, auxdata], [c])
# OCP boundary conditions
ocp_bcs = ca.Function('ocp_bcs', [xi, ui, xf, uf, p, auxdata], [b])
# OCP integral constraints
ocp_int = ca.Function('ocp_int', [t, x1, u, x2, p, h, auxdata], [dq])

# Build OCP
builder = minosPy.Builder(name)
builder.generate(ocp_runcost, ocp_bcscost, ocp_dyn, ocp_path, ocp_bcs, ocp_int, skip_hessian)
builder.build()

# Solve OCP
import numpy as npy
# create ocp object
ocp = minosPy.OCP(name, N, ti, tf) 
# get dims
dims = ocp.get_dims()
# set bounds
lbx = npy.array([0.0, 0.0, -50.0])
ubx = npy.array([10.0, 10.0, 50.0])
lbu = npy.array([-3.14/2])
ubu = npy.array([+3.14/2])
lbp = npy.array([0.0])
ubp = npy.array([5.0])
lbc = None
ubc = None
lbb = npy.array([0.0, 0.0, 0.0, 2.0, 2.0])
ubb = npy.array([0.0, 0.0, 0.0, 2.0, 2.0])
lbq = None
ubq = None
ocp.set_bounds(lbx, ubx, lbu, ubu, lbp, ubp, lbc, ubc, lbb, ubb, lbq, ubq)
# set guess
x0 = npy.vstack((npy.linspace(0, 2, N), 
                 npy.linspace(0, 2, N), 
                 npy.linspace(0, 0, N)))
u0 = npy.linspace(0, 1, N)
p0 = npy.array([1.0])
ocp.set_guess(x0, u0, p0)
# solver with different auxdata
auxdata_list = [npy.array([5.0]), 
                npy.array([10.0]), 
                npy.array([15.0])]
sol_list = []
for auxdata in auxdata_list:
    # set auxdata
    ocp.set_auxdata(auxdata)
    # call to solve
    ocp.solve()
    # get sol and append to sol_list
    sol = ocp.get_sol()
    sol_list.append(sol)
    mu_curr = ocp.get_mu_curr()
    # Set new guess
    ocp.set_guess(sol["x"], sol["u"], sol["p"],
                  sol["lamx"], sol["lamu"], sol["lamp"], 
                  sol["lamf"], sol["lamc"], sol["lamb"])
    ocp.set_option(mu_init=mu_curr)

# Print sol
k = 1
for sol in sol_list:
    print("Obj sol %d: %f" % (k, sol["objval"]))
    k += 1
